File: clients/python/kjson/parser.py
"""kJSON parser implementation."""

import re
import uuid
from typing import Any, Dict, List, Optional, Union
from .types import BigInt, Decimal128, Date


class JSONDecodeError(ValueError):
    """JSON decoding error."""
    
    def __init__(self, msg: str, doc: str, pos: int):
        """Initialize with error details."""
        super().__init__(msg)
        self.msg = msg
        self.doc = doc
        self.pos = pos
        
        # Calculate line and column
        lines = doc[:pos].split('\n')
        self.lineno = len(lines)
        self.colno = len(lines[-1]) + 1 if lines else 1


class Parser:
    """kJSON parser with JSON5 support."""
    
    def __init__(self, text: str):
        """Initialize parser with text."""
        self.text = text
        self.pos = 0
        self.length = len(text)
    
    def parse(self) -> Any:
        """Parse the JSON text."""
        self.skip_whitespace()
        result = self.parse_value()
        self.skip_whitespace()
        
        if self.pos < self.length:
            raise JSONDecodeError(
                f"Extra data after JSON value",
                self.text,
                self.pos
            )
        
        return result
    
    def current_char(self) -> Optional[str]:
        """Get current character."""
        if self.pos < self.length:
            return self.text[self.pos]
        return None
    
    def peek_char(self, offset: int = 0) -> Optional[str]:
        """Peek at character at offset."""
        pos = self.pos + offset
        if pos < self.length:
            return self.text[pos]
        return None
    
    def advance(self, count: int = 1):
        """Advance position."""
        self.pos = min(self.pos + count, self.length)
    
    def skip_whitespace(self):
        """Skip whitespace and comments."""
        while self.pos < self.length:
            ch = self.text[self.pos]
            
            if ch in ' \t\n\r':
                self.advance()
            elif ch == '/' and self.peek_char(1) == '/':
                # Line comment
                self.advance(2)
                while self.pos < self.length and self.text[self.pos] != '\n':
                    self.advance()
            elif ch == '/' and self.peek_char(1) == '*':
                # Block comment
                self.advance(2)
                while self.pos < self.length - 1:
                    if self.text[self.pos] == '*' and self.text[self.pos + 1] == '/':
                        self.advance(2)
                        break
                    self.advance()
            else:
                break
    
    def parse_value(self) -> Any:
        """Parse any JSON value."""
        self.skip_whitespace()
        
        if self.pos >= self.length:
            raise JSONDecodeError("Unexpected end of JSON input", self.text, self.pos)
        
        ch = self.text[self.pos]
        
        # null
        if ch == 'n':
            return self.parse_null()
        
        # boolean or unquoted literal
        elif ch == 't' or ch == 'f':
            # Try unquoted literal first (could be UUID)
            saved_pos = self.pos
            try:
                return self.parse_unquoted_literal()
            except:
                self.pos = saved_pos
                return self.parse_boolean()
        
        # string
        elif ch == '"' or ch == "'" or ch == '`':
            return self.parse_string()
        
        # array
        elif ch == '[':
            return self.parse_array()
        
        # object
        elif ch == '{':
            return self.parse_object()
        
        # number or date/uuid
        elif ch == '-' or ch.isdigit():
            # Need to look ahead to determine if this is a number or UUID/Date
            # UUIDs have format: 8-4-4-4-12 hex digits
            # Dates have format: YYYY-MM-DD
            saved_pos = self.pos
            
            # Try to detect UUID pattern (8 hex chars followed by -)
            is_uuid = False
            if ch.isdigit() or (ch >= 'a' and ch <= 'f'):
                # Count hex digits
                hex_count = 0
                temp_pos = self.pos
                while temp_pos < self.length and hex_count < 9:
                    c = self.text[temp_pos]
                    if (c >= '0' and c <= '9') or (c >= 'a' and c <= 'f') or (c >= 'A' and c <= 'F'):
                        hex_count += 1
                        temp_pos += 1
                    elif c == '-' and hex_count == 8:
                        is_uuid = True
                        break
                    else:
                        break
            
            # Try to detect date pattern (YYYY-MM-DD)
            is_date = False
            if not is_uuid and ch.isdigit():
                # Check for YYYY-MM pattern
                if self.pos + 4 < self.length and self.text[self.pos + 4] == '-':
                    is_date = True
            
            if is_uuid or is_date:
                try:
                    return self.parse_unquoted_literal()
                except:
                    self.pos = saved_pos
                    return self.parse_number()
            else:
                return self.parse_number()
        
        # unquoted literal (UUID, Date)
        else:
            return self.parse_unquoted_literal()
    
    def parse_null(self) -> None:
        """Parse null value."""
        if self.text[self.pos:self.pos+4] == 'null':
            self.advance(4)
            return None
        raise JSONDecodeError("Invalid null value", self.text, self.pos)
    
    def parse_boolean(self) -> bool:
        """Parse boolean value."""
        if self.text[self.pos:self.pos+4] == 'true':
            self.advance(4)
            return True
        elif self.text[self.pos:self.pos+5] == 'false':
            self.advance(5)
            return False
        raise JSONDecodeError("Invalid boolean value", self.text, self.pos)
    
    def parse_string(self) -> str:
        """Parse string value."""
        quote = self.text[self.pos]
        if quote not in '"\'`':
            raise JSONDecodeError("Expected string quote", self.text, self.pos)
        
        self.advance()  # Skip opening quote
        result = []
        
        while self.pos < self.length:
            ch = self.text[self.pos]
            
            if ch == quote:
                self.advance()  # Skip closing quote
                return ''.join(result)
            
            elif ch == '\\':
                self.advance()
                if self.pos >= self.length:
                    raise JSONDecodeError("Unterminated string escape", self.text, self.pos)
                
                escape_ch = self.text[self.pos]
                if escape_ch == '"':
                    result.append('"')
                elif escape_ch == "'":
                    result.append("'")
                elif escape_ch == '`':
                    result.append('`')
                elif escape_ch == '\\':
                    result.append('\\')
                elif escape_ch == '/':
                    result.append('/')
                elif escape_ch == 'b':
                    result.append('\b')
                elif escape_ch == 'f':
                    result.append('\f')
                elif escape_ch == 'n':
                    result.append('\n')
                elif escape_ch == 'r':
                    result.append('\r')
                elif escape_ch == 't':
                    result.append('\t')
                elif escape_ch == 'u':
                    # Unicode escape
                    self.advance()
                    if self.pos + 4 > self.length:
                        raise JSONDecodeError("Invalid unicode escape", self.text, self.pos)
                    hex_digits = self.text[self.pos:self.pos+4]
                    try:
                        code_point = int(hex_digits, 16)
                        result.append(chr(code_point))
                        self.advance(3)  # We already advanced 1
                    except ValueError:
                        raise JSONDecodeError("Invalid unicode escape", self.text, self.pos)
                else:
                    raise JSONDecodeError(f"Invalid escape sequence: \\{escape_ch}", self.text, self.pos)
                
                self.advance()
            
            else:
                result.append(ch)
                self.advance()
        
        raise JSONDecodeError("Unterminated string", self.text, self.pos)
    
    def parse_number(self) -> Union[int, float, BigInt, Decimal128]:
        """Parse number value."""
        start = self.pos
        
        # Optional negative
        if self.current_char() == '-':
            self.advance()
        
        # Integer part
        if self.current_char() == '0':
            self.advance()
        elif self.current_char() and self.current_char().isdigit():
            while self.current_char() and self.current_char().isdigit():
                self.advance()
        else:
            raise JSONDecodeError("Invalid number", self.text, self.pos)
        
        # Fractional part
        has_fraction = False
        if self.current_char() == '.':
            has_fraction = True
            self.advance()
            if not self.current_char() or not self.current_char().isdigit():
                raise JSONDecodeError("Invalid number: expected digits after decimal", self.text, self.pos)
            while self.current_char() and self.current_char().isdigit():
                self.advance()
        
        # Exponent part
        has_exponent = False
        ch = self.current_char()
        if ch and ch in 'eE':
            has_exponent = True
            self.advance()
            ch = self.current_char()
            if ch and ch in '+-':
                self.advance()
            if not self.current_char() or not self.current_char().isdigit():
                raise JSONDecodeError("Invalid number: expected digits in exponent", self.text, self.pos)
            while self.current_char() and self.current_char().isdigit():
                self.advance()
        
        # Check for BigInt suffix
        if self.current_char() == 'n':
            self.advance()
            if has_fraction or has_exponent:
                raise JSONDecodeError("BigInt cannot have fractional or exponent parts", self.text, start)
            return BigInt(self.text[start:self.pos-1])
        
        # Check for Decimal128 suffix
        if self.current_char() == 'm':
            self.advance()
            return Decimal128(self.text[start:self.pos-1])
        
        # Regular number
        num_str = self.text[start:self.pos]
        if has_fraction or has_exponent:
            return float(num_str)
        else:
            value = int(num_str)
            # Keep as int if it fits in Python int range
            return value
    
    def parse_array(self) -> List[Any]:
        """Parse array value."""
        if self.current_char() != '[':
            raise JSONDecodeError("Expected '['", self.text, self.pos)
        
        self.advance()  # Skip '['
        result = []
        
        self.skip_whitespace()
        
        # Empty array
        if self.current_char() == ']':
            self.advance()
            return result
        
        while True:
            # Parse value
            result.append(self.parse_value())
            self.skip_whitespace()
            
            ch = self.current_char()
            if ch == ',':
                self.advance()
                self.skip_whitespace()
                # Allow trailing comma
                if self.current_char() == ']':
                    self.advance()
                    return result
            elif ch == ']':
                self.advance()
                return result
            else:
                raise JSONDecodeError("Expected ',' or ']' in array", self.text, self.pos)
    
    def parse_object(self) -> Dict[str, Any]:
        """Parse object value."""
        if self.current_char() != '{':
            raise JSONDecodeError("Expected '{'", self.text, self.pos)
        
        self.advance()  # Skip '{'
        result = {}
        
        self.skip_whitespace()
        
        # Empty object
        if self.current_char() == '}':
            self.advance()
            return result
        
        while True:
            # Parse key
            self.skip_whitespace()
            
            ch = self.current_char()
            if ch == '"' or ch == "'" or ch == '`':
                # Quoted key
                key = self.parse_string()
            else:
                # Unquoted key (JSON5)
                key = self.parse_unquoted_key()
            
            self.skip_whitespace()
            
            # Expect colon
            if self.current_char() != ':':
                raise JSONDecodeError("Expected ':' after object key", self.text, self.pos)
            self.advance()
            
            # Parse value
            value = self.parse_value()
            result[key] = value
            
            self.skip_whitespace()
            
            ch = self.current_char()
            if ch == ',':
                self.advance()
                self.skip_whitespace()
                # Allow trailing comma
                if self.current_char() == '}':
                    self.advance()
                    return result
            elif ch == '}':
                self.advance()
                return result
            else:
                raise JSONDecodeError("Expected ',' or '}' in object", self.text, self.pos)
    
    def parse_unquoted_key(self) -> str:
        """Parse unquoted object key (JSON5)."""
        start = self.pos
        
        # For object keys, we need to check if it could be a literal value
        # Try to parse as unquoted literal first
        saved_pos = self.pos
        try:
            # Check if it's a valid literal (UUID/Date)
            literal = self.parse_unquoted_literal()
            # If we got here, it's a literal value, not a key
            # This shouldn't happen in valid JSON - reset and treat as key
            self.pos = saved_pos
        except:
            # Not a literal, continue as key
            self.pos = saved_pos
        
        # First character must be letter, underscore, or dollar
        ch = self.current_char()
        if not ch or not (ch.isalpha() or ch in '_$'):
            raise JSONDecodeError("Invalid unquoted key", self.text, self.pos)
        
        self.advance()
        
        # Subsequent characters
        while self.current_char() and (self.current_char().isalnum() or self.current_char() in '_$' or self.current_char() == '-'):
            self.advance()
        
        return self.text[start:self.pos]
    
    def parse_unquoted_literal(self) -> Union[uuid.UUID, Date]:
        """Parse unquoted literal (UUID or Date)."""
        start = self.pos
        
        # Read until delimiter
        while self.pos < self.length:
            ch = self.text[self.pos]
            if ch in ' \t\n\r,]}':
                break
            self.advance()
        
        literal = self.text[start:self.pos]
        
        # Try UUID
        try:
            return uuid.UUID(literal)
        except ValueError:
            pass
        
        # Try Date
        try:
            return Date(literal)
        except:
            pass
        
        raise JSONDecodeError(f"Invalid unquoted literal: {literal}", self.text, start)


def parse(text: str) -> Any:
    """Parse kJSON text."""
    parser = Parser(text)
    return parser.parse()

File: clients/rust/src/parser.rs
use crate::error::{Error, Result};
use crate::types::{BigInt, Date, Decimal128};
use crate::value::Value;
use std::collections::HashMap;
use uuid::Uuid;

/// Parser state
pub struct Parser<'a> {
    input: &'a str,
    position: usize,
}

/// Parse a kJSON string into a Value
pub fn parse(input: &str) -> Result<Value> {
    let mut parser = Parser { input, position: 0 };
    parser.skip_whitespace();
    let value = parser.parse_value()?;
    parser.skip_whitespace();
    if parser.position < parser.input.len() {
        return Err(Error::ParseError {
            position: parser.position,
            message: "Unexpected characters after value".to_string(),
        });
    }
    Ok(value)
}

impl<'a> Parser<'a> {
    /// Current character
    fn current(&self) -> Option<char> {
        self.input.chars().nth(self.position)
    }

    /// Peek at character without advancing
    fn peek(&self) -> Option<char> {
        self.current()
    }

    /// Advance position by one character
    fn advance(&mut self) {
        if self.position < self.input.len() {
            self.position += self.current().unwrap().len_utf8();
        }
    }

    /// Skip whitespace and comments
    fn skip_whitespace(&mut self) {
        while let Some(ch) = self.current() {
            match ch {
                ' ' | '\t' | '\n' | '\r' => self.advance(),
                '/' => {
                    let next_pos = self.position + 1;
                    if next_pos < self.input.len() {
                        let next_ch = self.input.chars().nth(next_pos);
                        match next_ch {
                            Some('/') => {
                                // Line comment
                                self.advance(); // Skip first /
                                self.advance(); // Skip second /
                                while let Some(c) = self.current() {
                                    self.advance();
                                    if c == '\n' {
                                        break;
                                    }
                                }
                            }
                            Some('*') => {
                                // Block comment
                                self.advance(); // Skip /
                                self.advance(); // Skip *
                                let mut star_seen = false;
                                while let Some(c) = self.current() {
                                    self.advance();
                                    if star_seen && c == '/' {
                                        break;
                                    }
                                    star_seen = c == '*';
                                }
                            }
                            _ => return,
                        }
                    } else {
                        return;
                    }
                }
                _ => return,
            }
        }
    }

    /// Parse any value
    fn parse_value(&mut self) -> Result<Value> {
        self.skip_whitespace();

        match self.peek() {
            None => Err(Error::UnexpectedEof),
            Some('n') => self.parse_null(),
            Some('t') | Some('f') => {
                // Could be boolean or UUID starting with 't' or 'f'
                if let Ok(literal) = self.try_parse_unquoted_literal() {
                    Ok(literal)
                } else {
                    self.parse_bool()
                }
            }
            Some('"') | Some('\'') | Some('`') => self.parse_string(),
            Some('[') => self.parse_array(),
            Some('{') => self.parse_object(),
            Some('-') | Some('0'..='9') => {
                // Could be number or date/UUID
                if let Ok(literal) = self.try_parse_unquoted_literal() {
                    Ok(literal)
                } else {
                    self.parse_number()
                }
            }
            Some(_) => self.parse_unquoted_literal(),
        }
    }

    /// Parse null
    fn parse_null(&mut self) -> Result<Value> {
        if self.input[self.position..].starts_with("null") {
            self.position += 4;
            Ok(Value::Null)
        } else {
            Err(Error::ParseError {
                position: self.position,
                message: "Invalid null value".to_string(),
            })
        }
    }

    /// Parse boolean
    fn parse_bool(&mut self) -> Result<Value> {
        if self.input[self.position..].starts_with("true") {
            self.position += 4;
            Ok(Value::Bool(true))
        } else if self.input[self.position..].starts_with("false") {
            self.position += 5;
            Ok(Value::Bool(false))
        } else {
            Err(Error::ParseError {
                position: self.position,
                message: "Invalid boolean value".to_string(),
            })
        }
    }

    /// Parse string
    fn parse_string(&mut self) -> Result<Value> {
        let quote_char = match self.current() {
            Some('"') | Some('\'') | Some('`') => self.current().unwrap(),
            _ => {
                return Err(Error::ParseError {
                    position: self.position,
                    message: "Expected quote character".to_string(),
                });
            }
        };
        self.advance(); // Skip opening quote

        let mut result = String::new();
        let mut escape = false;

        while let Some(ch) = self.current() {
            if escape {
                match ch {
                    '"' => result.push('"'),
                    '\'' => result.push('\''),
                    '`' => result.push('`'),
                    '\\' => result.push('\\'),
                    '/' => result.push('/'),
                    'b' => result.push('\u{0008}'),
                    'f' => result.push('\u{000C}'),
                    'n' => result.push('\n'),
                    'r' => result.push('\r'),
                    't' => result.push('\t'),
                    'u' => {
                        // Unicode escape
                        self.advance();
                        let mut hex = String::new();
                        for _ in 0..4 {
                            if let Some(ch) = self.current() {
                                hex.push(ch);
                                self.advance();
                            } else {
                                return Err(Error::ParseError {
                                    position: self.position,
                                    message: "Invalid unicode escape".to_string(),
                                });
                            }
                        }
                        let code_point = u32::from_str_radix(&hex, 16)
                            .map_err(|_| Error::ParseError {
                                position: self.position,
                                message: "Invalid unicode escape".to_string(),
                            })?;
                        if let Some(ch) = char::from_u32(code_point) {
                            result.push(ch);
                        } else {
                            return Err(Error::ParseError {
                                position: self.position,
                                message: "Invalid unicode code point".to_string(),
                            });
                        }
                        escape = false;
                        continue;
                    }
                    _ => {
                        return Err(Error::ParseError {
                            position: self.position,
                            message: format!("Invalid escape sequence: \\{}", ch),
                        })
                    }
                }
                escape = false;
                self.advance();
            } else if ch == '\\' {
                escape = true;
                self.advance();
            } else if ch == quote_char {
                self.advance();
                return Ok(Value::String(result));
            } else {
                result.push(ch);
                self.advance();
            }
        }

        Err(Error::ParseError {
            position: self.position,
            message: "Unterminated string".to_string(),
        })
    }

    /// Parse array
    fn parse_array(&mut self) -> Result<Value> {
        if self.current() != Some('[') {
            return Err(Error::ParseError {
                position: self.position,
                message: "Expected '['".to_string(),
            });
        }
        self.advance();

        let mut items = Vec::new();
        self.skip_whitespace();

        if self.current() == Some(']') {
            self.advance();
            return Ok(Value::Array(items));
        }

        loop {
            items.push(self.parse_value()?);
            self.skip_whitespace();

            match self.current() {
                Some(',') => {
                    self.advance();
                    self.skip_whitespace();
                    // Allow trailing comma
                    if self.current() == Some(']') {
                        self.advance();
                        break;
                    }
                }
                Some(']') => {
                    self.advance();
                    break;
                }
                _ => {
                    return Err(Error::ParseError {
                        position: self.position,
                        message: "Expected ',' or ']'".to_string(),
                    })
                }
            }
        }

        Ok(Value::Array(items))
    }

    /// Parse object
    fn parse_object(&mut self) -> Result<Value> {
        if self.current() != Some('{') {
            return Err(Error::ParseError {
                position: self.position,
                message: "Expected '{'".to_string(),
            });
        }
        self.advance();

        let mut map = HashMap::new();
        self.skip_whitespace();

        if self.current() == Some('}') {
            self.advance();
            return Ok(Value::Object(map));
        }

        loop {
            // Parse key
            self.skip_whitespace();
            let key = match self.current() {
                Some('"') | Some('\'') | Some('`') => {
                    // Quoted key
                    match self.parse_string()? {
                        Value::String(s) => s,
                        _ => unreachable!(),
                    }
                }
                _ => {
                    // Unquoted key (JSON5 style)
                    self.parse_unquoted_key()?
                }
            };

            self.skip_whitespace();
            if self.current() != Some(':') {
                return Err(Error::ParseError {
                    position: self.position,
                    message: "Expected ':' after key".to_string(),
                });
            }
            self.advance();

            // Parse value
            let value = self.parse_value()?;
            map.insert(key, value);

            self.skip_whitespace();
            match self.current() {
                Some(',') => {
                    self.advance();
                    self.skip_whitespace();
                    // Allow trailing comma
                    if self.current() == Some('}') {
                        self.advance();
                        break;
                    }
                }
                Some('}') => {
                    self.advance();
                    break;
                }
                _ => {
                    return Err(Error::ParseError {
                        position: self.position,
                        message: "Expected ',' or '}'".to_string(),
                    })
                }
            }
        }

        Ok(Value::Object(map))
    }

    /// Parse unquoted key (JSON5 style)
    fn parse_unquoted_key(&mut self) -> Result<String> {
        let start = self.position;

        // First character must be letter, underscore, or dollar sign
        match self.current() {
            Some(ch) if ch.is_alphabetic() || ch == '_' || ch == '$' => {
                self.advance();
            }
            _ => {
                return Err(Error::ParseError {
                    position: self.position,
                    message: "Invalid unquoted key".to_string(),
                })
            }
        }

        // Subsequent characters
        while let Some(ch) = self.current() {
            if ch.is_alphanumeric() || ch == '_' || ch == '$' {
                self.advance();
            } else {
                break;
            }
        }

        Ok(self.input[start..self.position].to_string())
    }

    /// Parse number (including BigInt and Decimal128)
    fn parse_number(&mut self) -> Result<Value> {
        let start = self.position;

        // Optional negative sign
        if self.current() == Some('-') {
            self.advance();
        }

        // Integer part
        if self.current() == Some('0') {
            self.advance();
        } else {
            while let Some(ch) = self.current() {
                if ch.is_ascii_digit() {
                    self.advance();
                } else {
                    break;
                }
            }
        }

        // Fractional part
        let has_decimal = self.current() == Some('.');
        if has_decimal {
            self.advance();
            let frac_start = self.position;
            while let Some(ch) = self.current() {
                if ch.is_ascii_digit() {
                    self.advance();
                } else {
                    break;
                }
            }
            if self.position == frac_start {
                return Err(Error::ParseError {
                    position: self.position,
                    message: "Expected digits after decimal point".to_string(),
                });
            }
        }

        // Exponent part
        let has_exponent = matches!(self.current(), Some('e') | Some('E'));
        if has_exponent {
            self.advance();
            if matches!(self.current(), Some('+') | Some('-')) {
                self.advance();
            }
            let exp_start = self.position;
            while let Some(ch) = self.current() {
                if ch.is_ascii_digit() {
                    self.advance();
                } else {
                    break;
                }
            }
            if self.position == exp_start {
                return Err(Error::ParseError {
                    position: self.position,
                    message: "Expected digits in exponent".to_string(),
                });
            }
        }

        // Check for BigInt suffix
        if self.current() == Some('n') {
            self.advance();
            let num_str = &self.input[start..self.position - 1];
            let bigint = BigInt::from_str(num_str)?;
            return Ok(Value::BigInt(bigint));
        }

        // Check for Decimal128 suffix
        if self.current() == Some('m') {
            self.advance();
            let num_str = &self.input[start..self.position - 1];
            let decimal = Decimal128::from_str(num_str)?;
            return Ok(Value::Decimal128(decimal));
        }

        // Regular number
        let num_str = &self.input[start..self.position];
        let num = num_str
            .parse::<f64>()
            .map_err(|_| Error::InvalidNumber(num_str.to_string()))?;
        Ok(Value::Number(num))
    }

    /// Try to parse unquoted literal (UUID, Date)
    fn try_parse_unquoted_literal(&mut self) -> Result<Value> {
        let saved_pos = self.position;
        match self.parse_unquoted_literal() {
            Ok(val) => Ok(val),
            Err(_) => {
                self.position = saved_pos;
                Err(Error::ParseError {
                    position: self.position,
                    message: "Not a valid literal".to_string(),
                })
            }
        }
    }

    /// Parse unquoted literal (UUID, Date)
    fn parse_unquoted_literal(&mut self) -> Result<Value> {
        let start = self.position;

        // Read until delimiter
        while let Some(ch) = self.current() {
            match ch {
                ' ' | '\t' | '\n' | '\r' | ',' | ']' | '}' => break,
                _ => self.advance(),
            }
        }

        let literal = &self.input[start..self.position];

        // Try to parse as UUID
        if let Ok(uuid) = Uuid::parse_str(literal) {
            return Ok(Value::Uuid(uuid));
        }

        // Try to parse as Date
        if let Ok(date) = Date::from_iso8601(literal) {
            return Ok(Value::Date(date));
        }

        Err(Error::ParseError {
            position: start,
            message: format!("Invalid literal: {}", literal),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_primitives() {
        assert_eq!(parse("null").unwrap(), Value::Null);
        assert_eq!(parse("true").unwrap(), Value::Bool(true));
        assert_eq!(parse("false").unwrap(), Value::Bool(false));
        assert_eq!(parse("123").unwrap(), Value::Number(123.0));
        assert_eq!(parse("3.14").unwrap(), Value::Number(3.14));
        assert_eq!(parse("\"hello\"").unwrap(), Value::String("hello".to_string()));
    }

    #[test]
    fn test_parse_extended_types() {
        // BigInt
        match parse("123456789012345678n").unwrap() {
            Value::BigInt(b) => assert_eq!(b.to_string(), "123456789012345678"),
            _ => panic!("Expected BigInt"),
        }

        // Decimal128
        match parse("99.99m").unwrap() {
            Value::Decimal128(d) => assert_eq!(d.to_string(), "99.99"),
            _ => panic!("Expected Decimal128"),
        }

        // UUID
        match parse("550e8400-e29b-41d4-a716-446655440000").unwrap() {
            Value::Uuid(u) => assert_eq!(u.to_string(), "550e8400-e29b-41d4-a716-446655440000"),
            _ => panic!("Expected UUID"),
        }

        // Date
        match parse("2025-01-10T12:00:00Z").unwrap() {
            Value::Date(_) => (), // Date parsing tested in types module
            _ => panic!("Expected Date"),
        }
    }

    #[test]
    fn test_parse_array() {
        let result = parse("[1, 2, 3]").unwrap();
        match result {
            Value::Array(arr) => {
                assert_eq!(arr.len(), 3);
                assert_eq!(arr[0], Value::Number(1.0));
                assert_eq!(arr[1], Value::Number(2.0));
                assert_eq!(arr[2], Value::Number(3.0));
            }
            _ => panic!("Expected array"),
        }
    }

    #[test]
    fn test_parse_object() {
        let result = parse(r#"{"name": "test", "value": 42}"#).unwrap();
        match result {
            Value::Object(obj) => {
                assert_eq!(obj.get("name"), Some(&Value::String("test".to_string())));
                assert_eq!(obj.get("value"), Some(&Value::Number(42.0)));
            }
            _ => panic!("Expected object"),
        }
    }

    #[test]
    fn test_parse_json5_features() {
        // Unquoted keys
        let result = parse("{name: \"test\", value: 42}").unwrap();
        match result {
            Value::Object(obj) => {
                assert_eq!(obj.get("name"), Some(&Value::String("test".to_string())));
                assert_eq!(obj.get("value"), Some(&Value::Number(42.0)));
            }
            _ => panic!("Expected object"),
        }

        // Trailing commas
        assert!(parse("[1, 2, 3,]").is_ok());
        assert!(parse("{a: 1,}").is_ok());
    }
}